nservicebus/azure-service-bus/performance-tuning_body_asb_[7,).partial.md

```diff
@@ -0,0 +1,63 @@
+
+  * `MessagingFactories().BatchFlushInterval()`: Controls the time that the ASB SDK batches client side requests before sending them out. Optimal values are between 100 and 500 milliseconds. To turn this feature off, provide TimeSpan.Zero.
+  * `MessagingFactories().NumberOfMessagingFactoriesPerNamespace()`: Each factory manages a TCP connection to a front end node in the broker, this connection has throughput limits. Opening more factories significantly improves send performance. Optimal values are between 2 and 8 for sending.
+  * `NumberOfClientsPerEntity()`: Keep this value equal to the number of factories, this ensures there is 1 send client per factory.
+
+The way the send API is invoked matters a lot as well. Using `await` on each `Send()` operation will make the ASB SDK wait until the send operation completes and therefore makes client side batching useless. It's better to maintain a list of send tasks and await them all together instead.
+
+Slow:
+
+Snippet: asb-slow-send
+
+Fast:
+
+Snippet: asb-fast-send
+
+## Tuning receive performance
+
+Tuning receive performance depends on quite a few things that vary per scenario, but the following settings bring improvements for all scenarios.
+
+  * `MessagingFactories().NumberOfMessagingFactoriesPerNamespace()`: Each factory manages a TCP connection to a front end node in the broker, this connection has throughput limits. Opening more factories significantly improves receive performance. Optimal value ares between 16 and 40 for receiving.
+  * `NumberOfClientsPerEntity()`: Keep this value equal to the number of factories, this ensures there is 1 internal receive client per factory.
+
+Other settings that impact performance a lot, but their values differ quite a bit per scenario, are:
+
+  * `LimitMessageProcessingConcurrencyTo()`: This is the global concurrency limit across all receive entities. Ensure this value matches the per receiver desired concurrency for the scenario, multiplied by the number of receivers.
+  * `MessageReceivers().PrefetchCount()`: Value is per receiver, in general the following rule applies: The lower the throughput rate of an individual receiver, the lower this value should be.
+  * `Transactions()`: The most significant performance impact comes from the transactional guarantees needed, `SendsAtomicWithReceive` will offer less performance then `ReceiveOnly`.
+
+### Sends Atomic With Receive
+
+In order to guarantee that sends occur as an atomic operation with the receive confirmation, the ASB SDK requires that both operations are enclosed in a `Serializable` transaction. This implies that:
+
+  * All atomic operations are executed in a 'Serial' fashion, i.e. 1 by 1, therefore it does not make sense to allow many concurrent operations per client, nor does it make much sense to prefetch a lot of messages.
+	  - Concurrency: Optimal values are 2 or 4 per client
+	  - PrefetchCount: Optimal value is 20
```

Comment:
the offical doco seems to suggest that using more receivers per factory works best, but this does not seem to be the case in our testing for this mode. and the prefetch value is also very exagerated it resulted in message lock lost exceptions for me. Now I don't know what mode was used by ASB team to come up with these values, they don't mention that.